homeassistant/components/mqtt/util.py

```diff
@@ -41,6 +53,57 @@ def mqtt_config_entry_enabled(hass: HomeAssistant) -> bool | None:
     return not bool(hass.config_entries.async_entries(DOMAIN)[0].disabled_by)
 
 
+async def async_wait_for_mqtt_client(hass: HomeAssistant) -> bool:
+    """Wait for the MQTT client to become available.
+
+    Waits when mqtt set up is in progress,
+    It is not needed that the client is connected.
+    Returns True if the mqtt client is available.
+    Returns False when the client is not available.
+    """
+    if not mqtt_config_entry_enabled(hass):
+        return False
+
+    entry = hass.config_entries.async_entries(DOMAIN)[0]
+    if entry.state == ConfigEntryState.LOADED:
+        return True
+
+    valid_states = {ConfigEntryState.SETUP_IN_PROGRESS, ConfigEntryState.LOADED}
+    state_reached_future: asyncio.Future[None] = asyncio.Future()
```

Comment:
I think you can make this simpler by storing the future in `DATA_MQTT_AVAILABLE` (or _READY)
and setting it to `True` when it sets up and `False` when it fails.  I'd also leave it there instead of cleaning up so you could replace it it somehow is successful later.

That you only need to await the future and get the result, and if setup fails or is successful you'll know

Reply:
I have now stored the `Future` and that makes it simpler. Only thing is to check if the `Future` is not `done`  before setting a result. I improved the tests with a simulation of some waiting subscribing integrations.